File: dart/test/shared.dart
String formatarErro(String str, String strR){
    return 'parametro(s):' + str + ' esperado -> ' + strR;
}

const _primitiveTypesPattern = ["int","double","num","String","bool"];

String? _isPrimitiveType(dynamic type){
  var _type = type.runtimeType.toString();
  if(_type.contains("List")){
    _type = _type.replaceAll("List", "").replaceAll("<", "").replaceAll(">", "");
  }
  if(_primitiveTypesPattern.contains(_type)){
    return _type;
  }
  return null;
}

bool _isMapStringDynamic(dynamic obj){
  return obj.runtimeType.toString() == "List<Map<String, dynamic>>";
}

bool equalsList(List<dynamic> lst1, List<dynamic> lst2){
  if(lst1.length != lst2.length) return false;
  var primitiveType1 = _isPrimitiveType(lst1);
  var primitiveType2 = _isPrimitiveType(lst2);
  if(primitiveType1 != primitiveType2) return false;

  var isPrimitive = primitiveType1 != null;
  var equals = false;
  for (var index = 0; index < lst1.length; index++) {
    equals = false;
    for (var index1 = 0; index1 < lst2.length; index1++) {
      if(isPrimitive){
        if(lst1[index] == lst2[index1]){
          equals = true;
          break;
        }
      }
      else if(!_isMapStringDynamic(lst1[index]) || !_isMapStringDynamic(lst2[index1])){
        equals = false;
        break;
      }
      else if(equalsObj(lst1[index], lst2[index1])){
        equals = true;
        break;
      }
    }
    if(!equals){
      break;
    }
  }
  return equals;
}

bool equalsObj(Map<String, dynamic> obj1, Map<String, dynamic> obj2){
    if(obj1.keys.length != obj2.keys.length) return false;
    var isEquals = true;
    for (var key in obj1.keys) {
        if(obj1[key] is Map && obj2[key] is Map){
          isEquals = equalsObj(obj1[key], obj2[key]);
          if(!isEquals){
            isEquals = false;
            break;
          }
        }
        else if(obj1[key] is List && obj2[key] is List){
          if(obj1[key].length != obj2[key].length){
            isEquals = false;
            break;
          }
          if(obj1[key].runtimeType.toString() != "List<Map<String, dynamic>>") return false;
          isEquals = equalsListObjs(obj1[key], obj2[key]);
          if(!isEquals){
            isEquals = false;
            break;
          }
        }
        else if(obj1[key] != obj2[key]){
          isEquals = false;
          break;
        }
    }
    return isEquals;
}

bool equalsListObjs(List<Map<String, dynamic>> lst1, List<Map<String, dynamic>> lst2){
  if(lst1.length != lst2.length) return false;
  var countEquals = 0;
  for (var map1 in lst1) {
    for (var map2 in lst2) {
      if(equalsObj(map1, map2)){
        countEquals++;
      }
    }
  }
  return countEquals == lst1.length;
}

File: CSharp/CSharp/Test/Shared.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace CSharp.Test
{
    public class Shared
    {
        public string FormatarErro(string str, string strR)
        {
            return "parametro(s):" + str + " esperado -> " + strR;
        }

        private bool _IsList(Type type)
        {
            return type.IsArray || type.Name.Contains("List");
        }

        public bool EqualsDictionary(Dictionary<string, string> d1, Dictionary<string, string> d2)
        {
            return d1.OrderBy(kvp => kvp.Key).SequenceEqual(d2.OrderBy(kvp => kvp.Key));
        }
        public bool EqualsDictionary(Dictionary<string, object[]> d1, Dictionary<string, object[]> d2)
        {
            if (d1.Count != d2.Count) return false;
            foreach (var item in d1)
            {
                if (!d2.ContainsKey(item.Key)) return false;
                var d2Value = d2[item.Key];
                if (!_EqualsListObjs(item.Value, d2Value)) return false;
            }
            return true;
        }

        private bool _IsPrimitiveType(Type type)
        {
            return
                type.IsPrimitive ||
                new Type[] {
                typeof(int),
                typeof(double),
                typeof(float),
                typeof(string),
                typeof(decimal),
                typeof(DateTime),
                typeof(DateTimeOffset),
                typeof(TimeSpan),
                typeof(Guid)
                }.Contains(type);
        }

        public bool EqualsObj(object obj1, object obj2)
        {
            if (obj1 == null || obj2 == null) return false;

            if (_IsPrimitiveType(obj1.GetType()))
            {
                if (obj1.ToString() != obj2.ToString())
                {
                    return false;
                }
                return true;
            }

            var propsObj1 = obj1.GetType().GetProperties().OrderBy(p => p.Name).ToArray();
            var propsObj2 = obj2.GetType().GetProperties().OrderBy(p => p.Name).ToArray();

            if (propsObj1.Length != propsObj2.Length) return false;
            var isEquals = true;
            for (var index = 0; index < propsObj1.Length; index++)
            {
                var prop1 = propsObj1[index];
                var prop2 = propsObj2[index];

                var val1 = prop1.GetValue(obj1);
                var val2 = prop2.GetValue(obj2);

                if (_IsList(val1.GetType()) && _IsList(val2.GetType()))
                {
                    if (val1.GetType().GetProperties().Length != val2.GetType().GetProperties().Length)
                    {
                        isEquals = false;
                        break;
                    }
                    isEquals = _EqualsListObjs(val1 as IList, val2 as IList);
                    if (!isEquals)
                    {
                        isEquals = false;
                        break;
                    }
                }
                else if (!_IsPrimitiveType(val1.GetType()))
                {
                    isEquals = EqualsObj(val1, val2);
                    if (!isEquals)
                    {
                        isEquals = false;
                        break;
                    }
                }
                else if (!val1.Equals(val2))
                {
                    isEquals = false;
                    break;
                }
            }
            return isEquals;
        }


        private bool _EqualsListObjs(IList lst1, IList lst2)
        {
            if (lst1.Count != lst2.Count) return false;
            var countEquals = 0;
            for (var index1 = 0; index1 < lst1.Count; index1++)
            {
                for (var index2 = 0; index2 < lst2.Count; index2++)
                {
                    if (EqualsObj(lst1[index1], lst2[index2]))
                    {
                        countEquals++;
                    }
                }
            }
            return countEquals == lst1.Count;
        }
    }
}